### PR TITLE
Zero-alloc trie lookups

### PR DESCRIPTION
**Pull request by rphmeier**
Trie half of #3993 

Also does some more code cleanup and adds an `OwnedNode` type which can be used for trie iterators.

The shared portion of lookup with TrieDB and TrieDBMut is refactored into a `lookup` module which was simpler to write iteratively. When HashDB gets `get_with` then it will use that and be truly zero-alloc.

### CODE REVIEW

File: util/src/trie/mod.rs [hunk 1]
```diff
@@ -76,6 +77,46 @@ pub type Result<T> = ::std::result::Result<T, Box<TrieError>>;
...
+
+	/// Decode a byte-slice into the desired item.
+	fn decode(self, &[u8]) -> Self::Item;
+
+	/// Record that a node has been passed through.
+	fn record(&mut self, &H256, &[u8], u32) { }
+}
+
+impl<'a> Query for &'a mut Recorder {
+	type Item = DBValue;
```
Comment (rphmeier) [inline util/src/trie/mod.rs]: Since recording queries often don't care about the value itself it might make sense just to use `()` here.

**Comment by gavofyork** [comment]
an additional test or two for the new API?

**Comment by rphmeier** [comment]
@gavofyork Not 100% sure what there _is_ left to test -- Regular `get` still works from the existing test corpus. Recording still works as expected from the recorder tests. The `Query` trait works because A) it compiles, and B) The `get` and `record` tests are using it.

**Comment by gavofyork** [comment]
didn't see any unit tests specifically targeting `get_with`

**Pull request merged**